File: ScanProject/JSAnalysis/SonarJS/Hard/1839-concatenated-words.js
/**https://leetcode.com/problems/concatenated-words/ */
//Given an array of strings words (without duplicates), return all the concatenated words in the given list of words.
//A concatenated word is defined as a string that is comprised entirely of at least two shorter words (not necesssarily distinct) in the given array.


/**
 * @param {string[]} words
 * @return {string[]}
 */
var findAllConcatenatedWordsInADict = function(words) {
    let n = words.length;
    let dp = [];
    for(let i = 0; i <= n; i++){
        dp.push([]);
        for(let j = 0; j <= n; j++){
            dp[i].push(false);
        }
    }
    for(let i = 1; i <= n; i++){
        for(let j = 1; j <= n; j++){
            if(i !== j){
                dp[i][j] = isConcatenated(words[i - 1], words[j - 1]);
            }
        }
    }
    let result = [];
    for(let i = 1; i <= n; i++){
        for(let j = 1; j <= n; j++){
            if(dp[i][j]){
                result.push(words[i - 1]);
            }
        }
    }
    return result;
    
    
    
};
function isConcatenated(s1, s2){
    if(s1.length > s2.length){
        return false;
    }
    if(s1.length === 0){
        return true;
    }
    let i = 0;
    let j = 0;
    while(i < s1.length && j < s2.length){
        if(s1[i] === s2[j]){
            if(isConcatenated(s1.substring(i + 1), s2.substring(j + 1))){
                return true;
            }
        }
        j++;
    }
    return false;
}
    

File: ScanProject/JavaAnalysis/SonarJava/src/main/java/Hard/1839-concatenated-words.java
package main.java.Hard;

import java.util.*;

/**https://leetcode.com/problems/concatenated-words/ */
//Given an array of strings words (without duplicates), return all the concatenated words in the given list of words.
//A concatenated word is defined as a string that is comprised entirely of at least two shorter words (not necesssarily distinct) in the given array.
// 
//Example 1:
//Input: words = ["cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"]
//Output: ["catsdogcats","dogcatsdog","ratcatdogcat"]
//Explanation: "catsdogcats" can be concatenated by "cats", "dog" and "cats"; 
//"dogcatsdog" can be concatenated by "dog", "cats" and "dog"; 
//"ratcatdogcat" can be concatenated by "rat", "cat", "dog" and "cat".
//Example 2:
//Input: words = ["cat","dog","catdog"]
//Output: ["catdog"]
// 
//Constraints:
//	1 <= words.length <= 104
//	1 <= words[i].length <= 30
//	words[i] consists of only lowercase English letters.
//	All the strings of words are unique.
//	1 <= sum(words[i].length) <= 105
class Solution11 {
    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        List<String> res = new ArrayList<>();
        Set<String> preWords = new HashSet<>();
        Arrays.sort(words, (a, b) -> (a.length() - b.length()));
        for(int i = 0; i < words.length; i++){
            if(canForm(words[i], preWords)){
                res.add(words[i]);
            }
            preWords.add(words[i]);
        }
        return res;
    }
    private boolean canForm(String word, Set<String> dict){
        if(dict.isEmpty()){
            return false;
        }
        boolean[] dp = new boolean[word.length() + 1];
        dp[0] = true;
        for(int i = 1; i <= word.length(); i++){
            for(int j = 0; j < i; j++){
                if(!dp[j]){
                    continue;
                }
                if(dict.contains(word.substring(j, i))){
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[word.length()];
    }
}